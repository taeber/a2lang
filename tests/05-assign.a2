asm {
	ORG $800
	JSR main
	JMP EXIT
}

use [
    EXIT  : sub @ $3D0
    INIT  : sub @ $FB2F  ; AKA TEXT
    HOME  : sub @ $FC58
    COUT  : sub <- [ch: char @ A] @ $FDED
    CROUT : sub @ $FD8E
    PRNTYX: sub <- [val: word @ YX] @ $F940
    PRBYTE: sub <- [byte: byte @ A] @ $FDDA
]

let Assert = sub <- [actual: byte, expected: byte] {
    PRBYTE(expected)
    if actual == expected {
        COUT(`=)
        COUT(`=)
        PRBYTE(actual)
        CROUT()
        ->
    }
    COUT(`<)
    COUT(`>)
    PRBYTE(actual)
    CROUT()
    COUT(`F)
    COUT(`A)
    COUT(`I)
    COUT(`L)
    COUT(`!)
    CROUT()
    EXIT()
}

let main = sub {

}

let NUMPEEPS = 10

var i: int
i := 0
while i < NUMPEEPS {
    i += 1
}
while i > 0 {
    i -= NUMPEEPS
}

let (
bool = :byte,
false = 0,
true = 1,
)

var [
    done: bool,
    working: bool,
]
done := false
; In C, this would be  working = !done
working != done
done := true

var flags: byte
flags := %00000000
flags |= %11111111  ; ORA
flags &= %10101010  ; AND
flags ^= %10101010  ; EOR

