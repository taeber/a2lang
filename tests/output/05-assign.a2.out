	ORG $800
	JSR main
	JMP EXIT
EXIT	EQU $3D0
INIT	EQU $FB2F
HOME	EQU $FC58
COUT	EQU $FDED
CROUT	EQU $FD8E
PRNTYX	EQU $F940
PRBYTE	EQU $FDDA
* COPYBB @A Assert.expected
Assert	LDA Assert.expected
	JSR PRBYTE
* IFEQ Assert.actual Assert.expected
*   Assert._1 Assert._2
Assert._0	LDA Assert.actual
	CMP Assert.expected
	BEQ Assert._1
	JMP Assert._2
* COPYBB @A #"="
Assert._1	LDA #"="
	JSR COUT
* COPYBB @A #"="
	LDA #"="
	JSR COUT
* COPYBB @A Assert.actual
	LDA Assert.actual
	JSR PRBYTE
	JMP CROUT
* COPYBB @A #"<"
Assert._2	LDA #"<"
	JSR COUT
* COPYBB @A #">"
	LDA #">"
	JSR COUT
* COPYBB @A Assert.actual
	LDA Assert.actual
	JSR PRBYTE
	JSR CROUT
* COPYBB @A #"F"
	LDA #"F"
	JSR COUT
* COPYBB @A #"A"
	LDA #"A"
	JSR COUT
* COPYBB @A #"I"
	LDA #"I"
	JSR COUT
* COPYBB @A #"L"
	LDA #"L"
	JSR COUT
* COPYBB @A #"!"
	LDA #"!"
	JSR COUT
	JSR CROUT
	JMP EXIT
main	RTS
NUMPEEPS	EQU $0A
* COPYBB i #$00
	LDA #$00
	STA i
* IFLT i #NUMPEEPS
*   A2_4 A2_5
A2_3	LDA i
	CMP #NUMPEEPS
	BCC A2_4
	JMP A2_5
* ADDBB i #$01
A2_4	LDA #$01
	CLC
	ADC i
	STA i
	JMP A2_3
* IFLT #$00 i
*   A2_6 A2_7
A2_5	LDA #$00
	CMP i
	BCC A2_6
	JMP A2_7
* SUBBB i #NUMPEEPS
A2_6	LDA #NUMPEEPS
	SEC
	SBC i
	STA i
	JMP A2_5
false	EQU $00
true	EQU $01
* COPYBB done #false
A2_7	LDA #false
	STA done
* TODO: handle STMT_ASSIGN for !=
* COPYBB done #true
	LDA #true
	STA done
* COPYBB flags #$00
	LDA #$00
	STA flags
* TODO: handle STMT_ASSIGN for |=
* TODO: handle STMT_ASSIGN for &=
* TODO: handle STMT_ASSIGN for ~=
Assert.actual	HEX 00
Assert.expected	HEX 00
i	HEX 00
done	HEX 00
working	HEX 00
flags	HEX 00
